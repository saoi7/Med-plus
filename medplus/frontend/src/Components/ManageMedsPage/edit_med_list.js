import React from 'react';
import NavBar from '../NavBar';
import { ImPlus } from 'react-icons/im';
import { Link, Switch, Route, BrowserRouter as Router } from 'react-router-dom';
import { withAuthorization } from '../Session';
import { withFirebase } from '../Firebase';
import * as ROUTES from '../../constants/routes';

class EditMedListBase extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            is_done_loading: false,
            firebase_error_flag: false,
            med_entries: []
        }
    }

    componentDidMount() {
        this.setState({ is_done_loading: false, firebase_error_flag: false, med_entries: [] });
        this.props.firebase.TEST_schedules().get().then(snapshot => {
            let db_meds_entries = snapshot.val();
            if(!db_meds_entries) {
                this.setState({ is_done_loading: true });
                return;
            }
            const med_items = Object.values(db_meds_entries);//.map(key => key);
            this.setState({ is_done_loading: true, med_entries: med_items });
        }).catch(err => {
            this.setState({ is_done_loading: true, firebase_error_flag: true });
        });
    }

    // TODO implement better views for loading and error conditions
    render() {
        let result = "loading...";
        if(this.state.is_done_loading && this.state.firebase_error_flag)
            result = "ERROR: failed to load data";
        if(this.state.is_done_loading && !this.state.firebase_error_flag) {
            result = this.state.med_entries.map((med_entry) => {
                return (
                    <Link className="no-underline" to={{ pathname: ROUTES.EDIT_MED, state: {...med_entry} }}>
                        <EditMedItem name={med_entry.med_name} />
                    </Link>
                );
            });
        }
        return (
            <div className="edit-container flex-container">
                { result }
            </div>
        );
    }
}

function EditMedItem(props) {
    return (
        <div className="link-button edit-button font-small">
            { props.name }
        </div>
    );
}

const EditMedList = withFirebase(EditMedListBase);
export default EditMedList
export { EditMedListBase, EditMedItem };
