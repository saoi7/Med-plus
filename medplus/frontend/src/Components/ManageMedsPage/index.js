import React from 'react';
import NavBar from '../NavBar';
import { ImPlus } from 'react-icons/im';
import { Link, Switch, Route, BrowserRouter as Router } from 'react-router-dom';
import { withAuthorization } from '../Session';
import { withFirebase } from '../Firebase';
import * as ROUTES from '../../constants/routes';
import EditMedList, { EditMedListBase, EditMedItem } from './edit_med_list';

// TODO update this so that it can dynamically resize the buttons to fit properly for larger numbers of medications

function ManageMedsPageBase(){
    return (
        <div className='background-with-logo-image edit-layout'>
            <div className="title font-large">
                Medication List
            </div>

            <EditMedList />

            <div className="add-new-button ">
                <Link to={ROUTES.ADD_MED} className="link-button edit-button font-small">
                    <div>Add new</div>
                    <ImPlus />
                </Link>
            </div>
            <NavBar />
        </div>
    );
}

const ManageMedsPage = withFirebase(ManageMedsPageBase);

const condition = authUser => !!authUser;
export default withAuthorization(condition)(ManageMedsPage);

export { ManageMedsPageBase, EditMedList, EditMedListBase, EditMedItem };
