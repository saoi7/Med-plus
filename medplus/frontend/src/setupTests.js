// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { unmountComponentAtNode } from "react-dom";

let test_dom_container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  test_dom_container = document.createElement("div");
  document.body.appendChild(test_dom_container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(test_dom_container);
  test_dom_container.remove();
  test_dom_container = null;
});

export { test_dom_container };
