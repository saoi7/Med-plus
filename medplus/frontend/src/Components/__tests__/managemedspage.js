import { MOCK_DB_MED_ENTRIES } from './mock_db_med_entries';
import {test_dom_container} from '../../setupTests';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from 'react-test-renderer';
import { getByTestId, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { MOCK_FIREBASE_PROP } from './mock_firebase_props';
import { EditMedListBase, EditMedItem } from '../ManageMedsPage/edit_med_list';
import flushPromises from 'flush-promises';

test("snapshot test of EditMedList", () => {
    const COMPONENT= renderer.create(
        <MemoryRouter>
            <EditMedListBase firebase={MOCK_FIREBASE_PROP} />
        </MemoryRouter>
    );
    let tree = COMPONENT.toJSON();
    expect(tree).toMatchSnapshot();
});

test("check number of meds listed in EditMedList is correct", async () => {
    act(() => {
        render(<MemoryRouter><EditMedListBase firebase={MOCK_FIREBASE_PROP} /></MemoryRouter>, test_dom_container);
    });
    await flushPromises();  // causes componentDidMount() to complete before continuing
    const LINKS = test_dom_container.childNodes[0].childNodes;

    expect(LINKS.length).toBe(Object.keys(MOCK_DB_MED_ENTRIES).length);
    const KEYS = Object.keys(MOCK_DB_MED_ENTRIES);
    const LINK_TEXT_CONTENT_ARR = Object.values(LINKS).map((link_element) => link_element.textContent );
    expect(LINK_TEXT_CONTENT_ARR.sort()).toEqual(KEYS.sort());
});

test("snapshot test of EditMedItem", () => {
    const NAME_PROP = 'abcdefg';
    const COMPONENT= renderer.create(
        <EditMedItem name={NAME_PROP} />
    );
    let tree = COMPONENT.toJSON();
    expect(tree).toMatchSnapshot();
});
