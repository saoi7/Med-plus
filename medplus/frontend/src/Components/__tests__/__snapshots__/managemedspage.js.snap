// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test of EditMedItem 1`] = `
<div
  className="link-button edit-button font-small"
>
  abcdefg
</div>
`;

exports[`snapshot test of EditMedList 1`] = `
<div
  className="edit-container flex-container"
>
  loading...
</div>
`;
